docs/ipv6-design-doc.md

```diff
@@ -0,0 +1,164 @@
+    fails.
+  * `use=if` does not filter out locally scoped or temporary IPv6
+    addresses.
+
+Some attempts have been made to add more robust IPv6 support:
+
+  * Debian's ddclient package applies a
+    [patch](https://salsa.debian.org/debian/ddclient/-/blob/67a138aa3d98d70f01766123f58ef40e98693fd4/debian/patches/usev6.diff)
+    that adds a new `usev6` option. The `usev6` option can be set to
+    `ip` or `if`, but not any of the other strategies currently
+    available for the `use` option (`web`, `cmd`, `fw`, `cisco`,
+    `cisco-asa`). When set to `ip` or `if`, only IPv6 addresses are
+    considered; IPv4 addresses are ignored. The patch does not change
+    the behavior of the `use` option, so `use=web` or `use=cmd` can be
+    used for IPv6 if pointed at something that only outputs an IPv6
+    address.
+  * [ddclient-curl](https://github.com/astlinux-project/ddclient-curl)
+    is a fork of ddclient that uses curl as the HTTP client (instead
+    of ddclient's own homemade client) for more robust IPv6 support.
+  * PR #40 is perhaps the most comprehensive attempt at adding full
+    IPv6 support, but it was never merged and has since
+    bit-rotted. There is renewed effort to rebase the changes and get
+    them merged in. PR #40 adds new options and changes some existing
+    options. The approach taken is to completely isolate IPv4 address
+    detection from IPv6 address detection and require the update
+    protocol callbacks to handle each type of address appropriately.
+
+## Requirements
+
+  * The mechanism for determining the current IPv4 address (the `use`
+    option) must be independently configurable from the mechanism used
+    to determine the current IPv6 address.
+  * The user must be able to disable IPv4 address updates without
+    affecting IPv6 updates.
+  * The user must be able to disable IPv6 address updates without
+    affecting IPv4 updates.
+  * If HTTP polling is used for both IPv4 and IPv6 address discovery,
+    the URL used to determine the IPv4 address (the `web` option) must
+    be independently configurable from the URL used to determine the
+    IPv6 address.
+  * The use of IPv4 or IPv6 to update a record must be independent of
+    the type of record being updated (IPv4 or IPv6).
+  * The callback for the update protocol must be given both addresses,
+    even if only one of the two addresses has changed.
+  * The callback for the update protocol must be told which addresses
+    have changed.
+
+### Nice-to-Haves
+
+  * The user should be able to force the update protocol to use IPv4
+    or IPv6.
+  * Unless configured otherwise, ddclient should first attempt to
+    update via IPv6 and fall back to IPv4 if the IPv6 connection
+    fails. This behavior can be added later; for now it is acceptable
+    to keep the current behavior (use IPv6 without IPv4 fallback if
+    there is a AAAA record, use IPv4 if there is no AAAA record).
```

Comment:
Why would we not leave this to the underlying network stack in the operating system?

Reply:
The recommended way to do "fallback" is [RFC 8305](https://tools.ietf.org/html/rfc8305). As far as I know there is no Perl module that implements that RFC so we would have to do it ourselves. Or we could use curl, which already implements RFC 8305 (or at least its predecessor, RFC 6555).